pretyping/nativenorm.ml

```diff
@@ -362,20 +362,24 @@ and nf_atom_type env sigma atom =
 
 
 and  nf_predicate env sigma ind mip params v pT =
-  match kind_of_value v, kind pT with
-  | Vfun f, Prod _ ->
+  match kind (whd_allnolet env pT) with
+  | LetIn (name,b,t,pT) ->
+      let dep,body =
+        nf_predicate (push_rel (LocalDef (name,b,t)) env) sigma ind mip params v pT in
+      dep, mkLetIn (name,b,t,body)
+  | Prod (name,dom,codom) -> begin
+    match kind_of_value v with
+    | Vfun f ->
       let k = nb_rel env in
       let vb = f (mk_rel_accu k) in
-      let name,dom,codom =
-	try decompose_prod env pT with
-	  DestKO ->
-	  CErrors.anomaly
-	    (Pp.strbrk "Returned a functional value in a type not recognized as a product type.")
-      in
       let dep,body = 
 	nf_predicate (push_rel (LocalAssum (name,dom)) env) sigma ind mip params vb codom in
       dep, mkLambda(name,dom,body)
-  | Vfun f, _ -> 
+    | _ -> assert false
```

Comment:
I'm a bit worried here, the previous function was not returning the same in the Prod/not Vfun case, right?

Reply:
Right. I implicitly made the assumption that the body was enough eta-expanded but maybe do I anticipate too much about the invariants to be preserved by CEP [#34](https://github.com/coq/ceps/pull/34). I can put a `nf_type` again with a note telling that this should not be necessary with a canonically eta-expanded form?

Reply:
In fact, I think there is a lot of legacy code around this `dep` thing. I'll remove it as part of another PR. For the scope of this PR, I suggest to put back the `nf_type` just in case.

Reply:
OK, done.